# Lesson 6 Class Exercises: More Practice

We will use today's class exercise to reinforce the learning we've done to this point

## Get Started
Import the Numpy and Pandas packages

In [1]:
import numpy as np
import pandas as pd

## Exercise 1

Read in the apple data found in the `data/apple_data.tsv` directory.

In [4]:
apple_data = pd.read_csv('../data/apple_data.tsv', sep="\t")
apple_data.head()

State    1982    1987    1992    1997    2002    2007     1982  \
0        Alabama    899     626     474     479     265     231    1,548    
1         Alaska      --      --      2       4       9      10        --   
2        Arizona    214     224     235     158     185     496    1,969    
3       Arkansas    512     451     247     238     164     153    2,276    
4     California  2,904   2,787   3,082   3,454   2,120   2,074   29,933    

      1987   1992.1   1997.1   2002.1   2007.1 2002.2 2007.2  2002.3  2007.3  
0   1,217    1,011      968      594      393     191    144     386     307  
1       --      (D)       2       13       15       8     10     (D)     (D)  
2   4,881    5,770    4,003    1,581    1,344     118    424   1,459   1,249  
3   2,120    1,430    1,048      658      287     123    116     560     220  
4  34,105   42,024   49,704   38,268   22,184   1,850  1,850  35,856  20,954

Explore the data using the approach we've used thus far:
- List all of the column names
- Identify the data types
- Find the shape of the data
- Explore categorical values -- do we have the expected number of values for each?
- Get descriptive statistics for the numerical data
- Do we have missing values?
- Do we have duplicates?

In [5]:
apple_data.columns

Index(['       State', '1982', '1987', '1992', '1997', '2002', '2007',
       '  1982', '  1987', '1992.1', '1997.1', '2002.1', '2007.1', '2002.2',
       '2007.2', '2002.3', '2007.3'],
      dtype='object')

In [6]:
apple_data.dtypes

       State    object
1982            object
1987            object
1992            object
1997            object
2002            object
2007            object
  1982          object
  1987          object
1992.1          object
1997.1          object
2002.1          object
2007.1          object
2002.2          object
2007.2          object
2002.3          object
2007.3          object
dtype: object

In [7]:
apple_data.shape

(50, 17)

In [8]:
apple_data.iloc[:,0].unique()

array(['   Alabama', '   Alaska', '   Arizona', '   Arkansas',
       '   California', '   Colorado', '   Connecticut', '   Delaware',
       '   Florida', '   Georgia', '   Hawaii', '   Idaho', '   Illinois',
       '   Indiana', '   Iowa', '   Kansas', '   Kentucky',
       '   Louisiana', '   Maine', '   Maryland', '   Massachusetts',
       '   Michigan', '   Minnesota', '   Mississippi', '   Missouri',
       '   Montana', '   Nebraska', '   Nevada', '   New Hampshire',
       '   New Jersey', '   New Mexico', '   New York',
       '   North Carolina', '   North Dakota', '   Ohio', '   Oklahoma',
       '   Oregon', '   Pennsylvania', '   Rhode Island',
       '   South Carolina', '   South Dakota', '   Tennessee', '   Texas',
       '   Utah', '   Vermont', '   Virginia', '   Washington',
       '   West Virginia', '   Wisconsin', '   Wyoming'], dtype=object)

In [9]:
apple_data.describe()

State 1982 1987  1992  1997  2002 2007   1982   1987 1992.1  \
count            50   50   50    50    50    50   50     50     50     50   
unique           50   49   48    49    49    50   49     49     49     49   
top         Alabama   --   --  392   485   265   35      --     --    (D)   
freq              1    2    2     2     2     1    2      2      2      2   

        1997.1 2002.1 2007.1 2002.2 2007.2 2002.3 2007.3  
count       50     50     50     50     50     50     50  
unique      48     48     48     46     49     46     48  
top     5,389     (D)    (D)      8    427    (D)    (D)  
freq         2      3      3      2      2      5      3

In [10]:
apple_data.isna().sum()

       State    0
1982            0
1987            0
1992            0
1997            0
2002            0
2007            0
  1982          0
  1987          0
1992.1          0
1997.1          0
2002.1          0
2007.1          0
2002.2          0
2007.2          0
2002.3          0
2007.3          0
dtype: int64

In [11]:
apple_data.duplicated().sum()

0

What needs cleaning with the data?

### Clean the data

Re-read the file and make sure we have get the missing values correct.

In [14]:
apple_data = pd.read_csv('../data/apple_data.tsv', sep="\t", na_values=['--', '(D)'])
apple_data.head()

State    1982    1987    1992    1997    2002    2007     1982  \
0        Alabama    899     626     474     479     265     231    1,548    
1         Alaska     NaN     NaN      2       4       9      10       NaN   
2        Arizona    214     224     235     158     185     496    1,969    
3       Arkansas    512     451     247     238     164     153    2,276    
4     California  2,904   2,787   3,082   3,454   2,120   2,074   29,933    

      1987   1992.1   1997.1   2002.1   2007.1 2002.2 2007.2  2002.3  2007.3  
0   1,217    1,011      968      594      393     191    144     386     307  
1      NaN      NaN       2       13       15       8     10     NaN     NaN  
2   4,881    5,770    4,003    1,581    1,344     118    424   1,459   1,249  
3   2,120    1,430    1,048      658      287     123    116     560     220  
4  34,105   42,024   49,704   38,268   22,184   1,850  1,850  35,856  20,954

Make the column names correct

In [15]:
apple_data.columns = ['State', 
                      'Farms-1982', 'Farms-1987', 'Farms-1992', 'Farms-1997', 'Farms-2002', 'Farms-2007',
                      'Acres-1982', 'Acres-1987', 'Acres-1992', 'Acres-1997', 'Acres-2002', 'Acres-2007', 
                      'Bearing_Farms-2002', 'Bearing_Farms-2007', 
                      'Bearing_Acres-2002', 'Bearing_Acres-2007']
apple_data.head()

State Farms-1982 Farms-1987 Farms-1992 Farms-1997 Farms-2002  \
0        Alabama       899        626        474        479        265    
1         Alaska        NaN        NaN         2          4          9    
2        Arizona       214        224        235        158        185    
3       Arkansas       512        451        247        238        164    
4     California     2,904      2,787      3,082      3,454      2,120    

  Farms-2007 Acres-1982 Acres-1987 Acres-1992 Acres-1997 Acres-2002  \
0       231      1,548      1,217      1,011        968        594    
1        10         NaN        NaN        NaN         2         13    
2       496      1,969      4,881      5,770      4,003      1,581    
3       153      2,276      2,120      1,430      1,048        658    
4     2,074     29,933     34,105     42,024     49,704     38,268    

  Acres-2007 Bearing_Farms-2002 Bearing_Farms-2007 Bearing_Acres-2002  \
0       393                 191                144                386   
1        15                   8                 10                NaN   
2     1,344                 118                424              1,459   
3       287                 123                116                560   
4    22,184               1,850              1,850             35,856   

  Bearing_Acres-2007  
0                307  
1                NaN  
2              1,249  
3                220  
4             20,954

Next, remove commas. We need to alter the strings of multiple columns. We know how to do this within a single column, but how do we do this with multiple columns?  Remember the `apply()` function! We can apply any function we want.  We can't use the `pd.Series.str.replace()` function though because that belongs to the Pandas Series objects, we need an independent function. We can write our own!

We can use the formal defintion (e.g.):

```python
def remoteCommas(col):
    return col.str.replace(',','')

apple_data = apple_data.apply(remoteCommas)
```

Or we can simplify it using Pythons `lambda` function definition. `lambda` is used to declare an in-line function.  It works just like a normal function but it does not have a name and is used for quick very short in-line functions.  For example:

```python
apple_data = apple_data.apply(lambda x: x.str.replace(',',''))
```

In [16]:
apple_data = apple_data.apply(lambda x: x.str.replace(',',''))
apple_data.head()

State Farms-1982 Farms-1987 Farms-1992 Farms-1997 Farms-2002  \
0        Alabama       899        626        474        479        265    
1         Alaska        NaN        NaN         2          4          9    
2        Arizona       214        224        235        158        185    
3       Arkansas       512        451        247        238        164    
4     California      2904       2787       3082       3454       2120    

  Farms-2007 Acres-1982 Acres-1987 Acres-1992 Acres-1997 Acres-2002  \
0       231       1548       1217       1011        968        594    
1        10         NaN        NaN        NaN         2         13    
2       496       1969       4881       5770       4003       1581    
3       153       2276       2120       1430       1048        658    
4      2074      29933      34105      42024      49704      38268    

  Acres-2007 Bearing_Farms-2002 Bearing_Farms-2007 Bearing_Acres-2002  \
0       393                 191                144                386   
1        15                   8                 10                NaN   
2      1344                 118                424               1459   
3       287                 123                116                560   
4     22184                1850               1850              35856   

  Bearing_Acres-2007  
0                307  
1                NaN  
2               1249  
3                220  
4              20954

Remove spaces preceeding the state names

In [17]:
apple_data['State'] = apple_data['State'].str.replace('   ','')
apple_data['State'].unique()

array(['Alabama', 'Alaska', 'Arizona', 'Arkansas', 'California',
       'Colorado', 'Connecticut', 'Delaware', 'Florida', 'Georgia',
       'Hawaii', 'Idaho', 'Illinois', 'Indiana', 'Iowa', 'Kansas',
       'Kentucky', 'Louisiana', 'Maine', 'Maryland', 'Massachusetts',
       'Michigan', 'Minnesota', 'Mississippi', 'Missouri', 'Montana',
       'Nebraska', 'Nevada', 'New Hampshire', 'New Jersey', 'New Mexico',
       'New York', 'North Carolina', 'North Dakota', 'Ohio', 'Oklahoma',
       'Oregon', 'Pennsylvania', 'Rhode Island', 'South Carolina',
       'South Dakota', 'Tennessee', 'Texas', 'Utah', 'Vermont',
       'Virginia', 'Washington', 'West Virginia', 'Wisconsin', 'Wyoming'],
      dtype=object)

Convert columns to numeric. Convert to floats instead of integers because NaNs are considered float values.

In [18]:
temp = apple_data.iloc[:,1:].apply(lambda x: x.astype('float'))
apple_data = pd.concat([apple_data['State'], temp], axis=1)
apple_data.head()

State  Farms-1982  Farms-1987  Farms-1992  Farms-1997  Farms-2002  \
0     Alabama       899.0       626.0       474.0       479.0       265.0   
1      Alaska         NaN         NaN         2.0         4.0         9.0   
2     Arizona       214.0       224.0       235.0       158.0       185.0   
3    Arkansas       512.0       451.0       247.0       238.0       164.0   
4  California      2904.0      2787.0      3082.0      3454.0      2120.0   

   Farms-2007  Acres-1982  Acres-1987  Acres-1992  Acres-1997  Acres-2002  \
0       231.0      1548.0      1217.0      1011.0       968.0       594.0   
1        10.0         NaN         NaN         NaN         2.0        13.0   
2       496.0      1969.0      4881.0      5770.0      4003.0      1581.0   
3       153.0      2276.0      2120.0      1430.0      1048.0       658.0   
4      2074.0     29933.0     34105.0     42024.0     49704.0     38268.0   

   Acres-2007  Bearing_Farms-2002  Bearing_Farms-2007  Bearing_Acres-2002  \
0       393.0               191.0               144.0               386.0   
1        15.0                 8.0                10.0                 NaN   
2      1344.0               118.0               424.0              1459.0   
3       287.0               123.0               116.0               560.0   
4     22184.0              1850.0              1850.0             35856.0   

   Bearing_Acres-2007  
0               307.0  
1                 NaN  
2              1249.0  
3               220.0  
4             20954.0

Check the types to make sure all is good

Now desribe the numeric data

### Explore the Data Again

Now that we have problems with the dataframe correct, let's explore. Are our data types correct?

In [19]:
apple_data.dtypes

State                  object
Farms-1982            float64
Farms-1987            float64
Farms-1992            float64
Farms-1997            float64
Farms-2002            float64
Farms-2007            float64
Acres-1982            float64
Acres-1987            float64
Acres-1992            float64
Acres-1997            float64
Acres-2002            float64
Acres-2007            float64
Bearing_Farms-2002    float64
Bearing_Farms-2007    float64
Bearing_Acres-2002    float64
Bearing_Acres-2007    float64
dtype: object

Now that we have numeric data we can perform a proper summary

In [20]:
apple_data.describe()

Farms-1982   Farms-1987   Farms-1992   Farms-1997   Farms-2002  \
count    48.00000    48.000000    49.000000    49.000000    50.000000   
mean    858.06250   764.958333   691.408163   690.510204   537.060000   
std     989.94688   915.886408   851.933974   912.523975   692.712221   
min      24.00000    18.000000     2.000000     4.000000     2.000000   
25%     237.00000   222.250000   235.000000   210.000000   168.750000   
50%     531.50000   544.500000   474.000000   428.000000   338.500000   
75%     932.00000   862.750000   782.000000   724.000000   578.250000   
max    5406.00000  5169.000000  4596.000000  4910.000000  3870.000000   

        Farms-2007     Acres-1982     Acres-1987     Acres-1992  \
count    50.000000      48.000000      48.000000      47.000000   
mean    511.820000   12302.979167   12521.354167   12416.255319   
std     606.840183   25304.382357   27199.429306   28246.151550   
min       9.000000      41.000000      33.000000      35.000000   
25%     148.500000    1383.000000    1170.000000     999.000000   
50%     333.000000    4086.500000    4536.500000    4332.000000   
75%     562.500000    8642.000000    8791.250000    7407.000000   
max    3052.000000  145630.000000  162609.000000  169107.000000   

          Acres-1997     Acres-2002     Acres-2007  Bearing_Farms-2002  \
count      47.000000      47.000000      47.000000           50.000000   
mean    12927.510638    9863.234043    8470.638298          421.680000   
std     34141.187355   27100.575228   25504.573629          613.202348   
min         2.000000       6.000000      15.000000            2.000000   
25%       973.500000     596.500000     373.000000          119.250000   
50%      3187.000000    2384.000000    2057.000000          232.500000   
75%      6465.500000    4202.500000    3966.000000          400.750000   
max    215463.000000  172810.000000  165215.000000         3580.000000   

       Bearing_Farms-2007  Bearing_Acres-2002  Bearing_Acres-2007  
count           50.000000           45.000000           47.000000  
mean           434.320000         8917.000000         7651.595745  
std            557.364954        24400.441667        23454.296203  
min              8.000000           19.000000            3.000000  
25%            116.500000          470.000000          261.500000  
50%            264.000000         2113.000000         1769.000000  
75%            453.250000         4083.000000         3444.000000  
max           2912.000000       152241.000000       152334.000000

Do we have any missing values?

In [21]:
apple_data.isna().sum()

State                 0
Farms-1982            2
Farms-1987            2
Farms-1992            1
Farms-1997            1
Farms-2002            0
Farms-2007            0
Acres-1982            2
Acres-1987            2
Acres-1992            3
Acres-1997            3
Acres-2002            3
Acres-2007            3
Bearing_Farms-2002    0
Bearing_Farms-2007    0
Bearing_Acres-2002    5
Bearing_Acres-2007    3
dtype: int64

Are missing values reasonable?  Should we remove any?

Do we have any duplicated data?

In [22]:
apple_data.duplicated().sum()

0

Does our only categorical column have expected values?

In [23]:
apple_data['State'].unique()

array(['Alabama', 'Alaska', 'Arizona', 'Arkansas', 'California',
       'Colorado', 'Connecticut', 'Delaware', 'Florida', 'Georgia',
       'Hawaii', 'Idaho', 'Illinois', 'Indiana', 'Iowa', 'Kansas',
       'Kentucky', 'Louisiana', 'Maine', 'Maryland', 'Massachusetts',
       'Michigan', 'Minnesota', 'Mississippi', 'Missouri', 'Montana',
       'Nebraska', 'Nevada', 'New Hampshire', 'New Jersey', 'New Mexico',
       'New York', 'North Carolina', 'North Dakota', 'Ohio', 'Oklahoma',
       'Oregon', 'Pennsylvania', 'Rhode Island', 'South Carolina',
       'South Dakota', 'Tennessee', 'Texas', 'Utah', 'Vermont',
       'Virginia', 'Washington', 'West Virginia', 'Wisconsin', 'Wyoming'],
      dtype=object)

How many rows do we have per state? 

In [24]:
apple_data.groupby('State').size()

State
Alabama           1
Alaska            1
Arizona           1
Arkansas          1
California        1
Colorado          1
Connecticut       1
Delaware          1
Florida           1
Georgia           1
Hawaii            1
Idaho             1
Illinois          1
Indiana           1
Iowa              1
Kansas            1
Kentucky          1
Louisiana         1
Maine             1
Maryland          1
Massachusetts     1
Michigan          1
Minnesota         1
Mississippi       1
Missouri          1
Montana           1
Nebraska          1
Nevada            1
New Hampshire     1
New Jersey        1
New Mexico        1
New York          1
North Carolina    1
North Dakota      1
Ohio              1
Oklahoma          1
Oregon            1
Pennsylvania      1
Rhode Island      1
South Carolina    1
South Dakota      1
Tennessee         1
Texas             1
Utah              1
Vermont           1
Virginia          1
Washington        1
West Virginia     1
Wisconsin         1
Wyoming       

Are the number of rows per state what we expect?

### Tidy the data

What is/are the observable unit(s)?

What are the variables?

What are the observations (values)?

Next, tidy the data by melting the data. We have multiple variables in the header and we must melt them individually.  The best way to do this is to melt each variable separately, then combine them back to a single data frame. 

Do this, you will need to perform multiple melts:
1. First melt only the colums that have the number of farms and name the column `Farms`.
2. Second, melt only the columns that have the number of acres and name the column `Acres`.
3. Third, melt only the columns that have the number of bearing farms and name the column `Bearing_Farms`,
4. Fourth, melt only the columns that have the number of bearing acres and name the column `Bearing_Acres`.

You'll also want to remove the prefix (e.g. "Farms-") from each value after melting and convert the 'Year' column to an integer.

For example, after the first melt you should have a dataframe that looks like the following:

|  | State | Year | Farms |
| - | ----- | ---- | --------- |
| 0 | Alabama | 1982 | 899 |
| 1 | Alaska | 1982 | NaN |
| 2 | Arizona | 1982 | 214 |
| 3 | Arkansas | 1982 | 512 |
| 4 | California | 1982 | 2904 |

Melt #1

In [25]:
# Melt the data
cols = ['State', 'Farms-1982', 'Farms-1987', 'Farms-1992', 'Farms-1997', 'Farms-2002', 'Farms-2007']
farms = apple_data[cols].melt(id_vars = 'State', var_name='Year', value_name='Farms')

# Remove the 'Farms-' string and convert the year to an int
farms['Year'] = farms['Year'].str.replace('Farms-','').astype(int)

print(farms.shape)
farms.head()

(300, 3)


State  Year   Farms
0     Alabama  1982   899.0
1      Alaska  1982     NaN
2     Arizona  1982   214.0
3    Arkansas  1982   512.0
4  California  1982  2904.0

Melt #2

In [26]:
cols = ['State', 'Acres-1982', 'Acres-1987', 'Acres-1992', 'Acres-1997', 'Acres-2002', 'Acres-2007', ]
acres = apple_data[cols].melt(id_vars = 'State', var_name='Year', value_name='Acres')
acres['Year'] = acres['Year'].str.replace('Acres-','').astype(int)
print(acres.shape)
acres.head()

(300, 3)


State  Year    Acres
0     Alabama  1982   1548.0
1      Alaska  1982      NaN
2     Arizona  1982   1969.0
3    Arkansas  1982   2276.0
4  California  1982  29933.0

Melt #3

In [27]:
cols = ['State', 'Bearing_Farms-2002', 'Bearing_Farms-2007']
bearing_farms = apple_data[cols].melt(id_vars = 'State', var_name='Year', 
                                      value_name='Bearing_Farms')
bearing_farms['Year'] = bearing_farms['Year'].str.replace('Bearing_Farms-','').astype(int)
print(bearing_farms.shape)
bearing_farms.head()

(100, 3)


State  Year  Bearing_Farms
0     Alabama  2002          191.0
1      Alaska  2002            8.0
2     Arizona  2002          118.0
3    Arkansas  2002          123.0
4  California  2002         1850.0

Melt #4

In [28]:
cols = ['State', 'Bearing_Acres-2002', 'Bearing_Acres-2007']
bearing_acres = apple_data[cols].melt(id_vars = 'State', var_name='Year', 
                                      value_name='Bearing_Acres')
bearing_acres['Year'] = bearing_acres['Year'].str.replace('Bearing_Acres-','').astype(int)
print(bearing_acres.shape)
bearing_acres.head()

(100, 3)


State  Year  Bearing_Acres
0     Alabama  2002          386.0
1      Alaska  2002            NaN
2     Arizona  2002         1459.0
3    Arkansas  2002          560.0
4  California  2002        35856.0

Now combine all four melted dataframes into a single dataframe. To do this we can use the [pd.DataFrame.join()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.join.html) function. We will need to join one table at a time.  

Note: when you join tables they must have the same indexes. So, be sure to set the index of each dataframe before joining by using the [set_index()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.set_index.html) function.

In [29]:
# Set the indexes to be the same
farms.set_index(['State','Year'], inplace=True)
acres.set_index(['State','Year'], inplace=True)
bearing_farms.set_index(['State','Year'], inplace=True)
bearing_acres.set_index(['State','Year'], inplace=True)

In [30]:
# Join all of the tables.
join1 = farms.join(acres)
join2 = join1.join(bearing_farms)
adm = join2.join(bearing_acres)
adm.head()

Farms    Acres  Bearing_Farms  Bearing_Acres
State      Year                                               
Alabama    1982   899.0   1548.0            NaN            NaN
Alaska     1982     NaN      NaN            NaN            NaN
Arizona    1982   214.0   1969.0            NaN            NaN
Arkansas   1982   512.0   2276.0            NaN            NaN
California 1982  2904.0  29933.0            NaN            NaN

What's the shape of the data now?

In [31]:
adm.shape

(300, 4)

Do we have any duplicated rows now?

In [32]:
adm.duplicated().sum()

5

In [33]:
adm[adm.duplicated()]

Farms  Acres  Bearing_Farms  Bearing_Acres
State  Year                                            
Hawaii 1982    NaN    NaN            NaN            NaN
Alaska 1987    NaN    NaN            NaN            NaN
Hawaii 1987    NaN    NaN            NaN            NaN
       1992    NaN    NaN            NaN            NaN
       1997    NaN    NaN            NaN            NaN

Now, do we want to remove missing values?

In [34]:
adm.dropna(how='all', inplace=True)
adm.shape

(294, 4)

### Explore the Data Again

Now that we have our data fully cleaned and tidy'ed let's explore it to make sure all looks good. First, we should still have 50 states. Can you find out?

In [35]:
adm.reset_index()['State'].unique().shape

(50,)

How many years do we have per state? We should have the same number except for the states where we removed rows with missing values

In [36]:
adm.groupby('State').size()

State
Alabama           6
Alaska            4
Arizona           6
Arkansas          6
California        6
Colorado          6
Connecticut       6
Delaware          6
Florida           6
Georgia           6
Hawaii            2
Idaho             6
Illinois          6
Indiana           6
Iowa              6
Kansas            6
Kentucky          6
Louisiana         6
Maine             6
Maryland          6
Massachusetts     6
Michigan          6
Minnesota         6
Mississippi       6
Missouri          6
Montana           6
Nebraska          6
Nevada            6
New Hampshire     6
New Jersey        6
New Mexico        6
New York          6
North Carolina    6
North Dakota      6
Ohio              6
Oklahoma          6
Oregon            6
Pennsylvania      6
Rhode Island      6
South Carolina    6
South Dakota      6
Tennessee         6
Texas             6
Utah              6
Vermont           6
Virginia          6
Washington        6
West Virginia     6
Wisconsin         6
Wyoming       

Describe the data, but make sure it does so by state.

In [37]:
adm.groupby('State').describe()

Farms                                                    \
               count         mean         std     min      25%     50%   
State                                                                    
Alabama          6.0   495.666667  246.441609   231.0   317.25   476.5   
Alaska           4.0     6.250000    3.862210     2.0     3.50     6.5   
Arizona          6.0   252.000000  122.782735   158.0   192.25   219.0   
Arkansas         6.0   294.166667  151.194466   153.0   182.50   242.5   
California       6.0  2736.833333  544.749637  2074.0  2286.75  2845.5   
Colorado         6.0   511.166667   91.759287   361.0   474.25   517.5   
Connecticut      6.0   268.833333   28.067181   242.0   247.25   260.0   
Delaware         6.0    19.333333    3.326660    16.0    17.25    18.0   
Florida          6.0   116.166667   51.689135    53.0    83.75   109.0   
Georgia          6.0   517.666667  232.708974   286.0   335.75   476.5   
Hawaii           2.0     5.500000    4.949747     2.0     3.75     5.5   
Idaho            6.0   356.666667   37.119626   315.0   326.25   358.0   
Illinois         6.0   679.666667  181.313724   444.0   538.50   703.5   
Indiana          6.0   584.000000  195.444110   265.0   485.25   650.0   
Iowa             6.0   416.333333   49.122975   344.0   382.75   441.0   
Kansas           6.0   266.333333   94.730495   147.0   209.25   262.5   
Kentucky         6.0   794.166667  168.466515   616.0   644.75   794.0   
Louisiana        6.0    87.500000   40.825237    35.0    58.50    87.5   
Maine            6.0   373.666667   22.277043   345.0   355.25   380.0   
Maryland         6.0   379.000000  110.742946   252.0   300.50   365.0   
Massachusetts    6.0   426.833333   70.703371   338.0   376.50   425.5   
Michigan         6.0  2264.833333  421.590520  1750.0  1901.50  2381.5   
Minnesota        6.0   478.500000   69.000725   391.0   427.25   482.0   
Mississippi      6.0   241.833333  114.251331    80.0   181.25   253.0   
Missouri         6.0   588.333333  159.323152   438.0   494.75   531.5   
Montana          6.0   161.333333   20.432001   142.0   145.75   154.0   
Nebraska         6.0   114.666667   11.587349   100.0   106.25   115.5   
Nevada           6.0    46.333333    6.918574    35.0    45.00    46.0   
New Hampshire    6.0   210.500000   14.989997   183.0   209.25   212.5   
New Jersey       6.0   512.500000   32.519225   471.0   491.00   515.5   
New Mexico       6.0   779.000000   81.422356   666.0   742.00   777.0   
New York         6.0  1618.833333  248.602025  1350.0  1474.50  1567.5   
North Carolina   6.0   982.666667  332.874551   598.0   742.00   969.5   
North Dakota     6.0    25.500000   11.467345     5.0    22.75    28.5   
Ohio             6.0  1435.166667  291.232153  1000.0  1312.50  1429.0   
Oklahoma         6.0   418.833333  196.986717   178.0   277.25   409.5   
Oregon           6.0  1683.166667  251.918572  1330.0  1534.75  1723.5   
Pennsylvania     6.0  2091.666667  308.678905  1777.0  1862.00  2034.5   
Rhode Island     6.0    63.833333    8.658329    51.0    58.75    64.5   
South Carolina   6.0   312.833333   61.745985   205.0   299.25   325.5   
South Dakota     6.0    47.000000   10.917875    35.0    39.25    44.0   
Tennessee        6.0   868.833333  355.561762   357.0   638.25   963.5   
Texas            6.0   711.166667  167.958824   437.0   628.00   753.0   
Utah             6.0   511.500000   87.142986   388.0   464.00   510.5   
Vermont          6.0   242.500000   15.591664   217.0   238.50   243.0   
Virginia         6.0   993.166667  280.663797   626.0   802.00  1013.5   
Washington       6.0  4500.500000  887.616528  3052.0  4051.50  4753.0   
West Virginia    6.0   558.833333   91.269747   463.0   496.75   545.5   
Wisconsin        6.0   945.833333   66.847339   856.0   898.25   950.0   
Wyoming          6.0    18.833333    3.816630    14.0    16.50    19.0   

                                Acres                 ... Bearing_Farms  \
                 